code/game/machinery/autolathe.dm

```diff
@@ -332,4 +299,97 @@
 		over_location = get_turf(over_location)
 	if(Adjacent(over_location) && Adjacent(usr))
 		for(var/obj/item/O in over_location)
-			attempt_fill(O, usr)
\ No newline at end of file
+			attempt_fill(O, usr)
+
+/obj/machinery/autolathe/proc/addToQueue(var/datum/autolathe/recipe/D)
+	queue += D
+	return
+
+/obj/machinery/autolathe/proc/removeFromQueue(var/index)
+	queue.Cut(index, index + 1)
+	return
+
+/obj/machinery/autolathe/Process()
+	if(queue.len == 0)
+		return
+	if(busy)
+		return
+	var/datum/autolathe/recipe/D = queue[1]
+	if(queue.len)
+		build(D)
+
+/obj/machinery/autolathe/proc/build(var/datum/autolathe/recipe/D)
+	busy = 1
+	update_use_power(2)
+	var/datum/autolathe/recipe/making
+	making = D
+
+	//Check if we still have the materials.
+	for(var/material in making.resources)
+		if(!isnull(stored_material[material]))
+			if(stored_material[material] < round(making.resources[material] * mat_efficiency))
+				visible_message("<span class='warning'>[src] buzzes, 'Not enough materials for [making.name]!</span>")
+				removeFromQueue(1)
+				return TOPIC_REFRESH
+
+	//Consume materials.
+	for(var/material in making.resources)
+		if(!isnull(stored_material[material]))
+			stored_material[material] = max(0, stored_material[material] - round(making.resources[material] * mat_efficiency))
+
+	//Fancy autolathe animation.
+	flick("autolathe_n", src)
+
+	sleep(build_time)
+
+	busy = 0
+	update_use_power(1)
+
+	//Sanity check.
+	if(!making || QDELETED(src)) return TOPIC_HANDLED
+	//Create the desired item.
+	new making.path(loc)
+	removeFromQueue(1)
+	busy = 0
+
+/obj/machinery/autolathe/verb/extract_materials(var/mob/living/carbon/human/user)
+	set name = "Extract Materials"
+	set category = "Object"
+	set src in view(1)
+
+	if(user.incapacitated() || !istype(user, /mob/living))
+		to_chat(user, "<span class='warning'>You can't do that.</span>")
+		return
+
+	if(!Adjacent(user))
+		to_chat(user, "<span class='warning'>You can't reach it.</span>")
+		return
+	var/extraction_choices = list("steel", "glass", "wood")
+	var/material = input(user, "What do you want to extract?", "Materials Extraction") as null|anything in extraction_choices
+	var/mat_number = input(user, "How much do you want to extract? (Max: 50)", "Materials Extraction") as num
+	mat_number = Clamp(mat_number, 0, 50)
+	var/extraction_path
+	switch(material)
+		if("steel")
+			extraction_path = /obj/item/stack/material/steel
+		if("glass")
+			extraction_path = /obj/item/stack/material/glass
+		if("wood")
+			extraction_path = /obj/item/stack/material/wood
+	//We have all the info we need now - calculate how much it'll cost.
+	var/mat_cost = mat_number * SHEET_MATERIAL_AMOUNT
+	var/stored_materials = stored_material[material]
+	if(stored_materials < mat_cost) //Not enough material stored. Settle for the next best thing.
```

Comment:
This isn't enough to prevent removing more than what exists in storage. If you put in 2 sheets, built something, and removed those 2 sheets you will get this. 
![image](https://user-images.githubusercontent.com/16085952/56754613-7ab6dc80-675b-11e9-94f9-a693cf669f9b.png)